### DISCUSSION TITLE
Pseudocode block filter

### DISCUSSION THREAD

**Discussion by camilogarciabotero** [quarto-dev/quarto-cli#3049] [Feature Requests]
Hi all,

I recently saw a lovely [template](https://github.com/computorg/computo-quarto-extension) for creating quarto articles and noticed that they have an excellent filter for creating [pseudocode blocks](https://github.com/computorg/computo-quarto-extension/blob/main/_extensions/computo/pseudocode.lua)

Which works basically like this:

<img width="634" alt="Screen Shot 2022-10-26 at 7 50 24 AM" src="https://user-images.githubusercontent.com/25586362/198030345-585201b7-9486-4ad0-973c-e702d6e6c744.png">

Here the rendered [template](https://computo.sfds.asso.fr/computo-quarto-extension/) 

So, I thought it might be a nice native quarto filter for having by default since it is only enabled in the template.

**Comment by dragonstyle** [comment]
Great suggestion! We spent some time looking at this but had trouble reconciling pseudocode for LaTeX output and for HTML - (e.g. there was variant syntax between `pseudocode.js` for HTML and the various algorithm packages for LaTeX. I spent some time trying to find a configuration that would result in a single input producing output in both formats but couldn't make it work - if anyone has figured that you (e.g. a configuration of `algorithmx` that will consume the same input that `pseudocode.js`), I would be very interested in making this work either as an extension or as a native Quarto feature.

**Reply by camilogarciabotero** [reply]
Hi Charles,

Well I've no experience with lua at all. This is the function they implemented:

```
function CodeBlock(cb)
    if cb.attr.classes[1] == "pseudocode" then
        if quarto.doc.is_format("pdf") then
            quarto.doc.use_latex_package("algorithm") 
            quarto.doc.use_latex_package("algpseudocodex")
            return pandoc.RawBlock('tex', cb.text)
        end
        if quarto.doc.is_format("html:js") then
            quarto.doc.addHtmlDependency({
                name = "pseudocode",
                version = "1.0.0",
                scripts = { "pseudocode.min.js" },
                stylesheets = { "pseudocode.min.css" }
            })
            -- generate the initialization script with the correct options
            local scriptTag = [[<script>
for (const element of document.getElementsByClassName("pseudocode")){
    pseudocode.renderElement(element);
}
</script>]]

            -- inject the rendering code
            quarto.doc.include_text("after-body", scriptTag)
        end
    end
end
```

It is linked in the comment above, but what I see is that they are using two $\LaTeX$ packages: `algorithm` and `algpseudocodex` that are called when the format is called pdf, whereas when the format is called HTML they send the call to `pseudocode.min.js` and a related CSS template `pseudocode.min.css` both present in the extension folder (probably you already know this... apologize).

It called my attention that when formatting the `_extension.yml` they used `html-math-method: katex` instruction instead of the default (which If I recall is actually `mathjax`). I don't really know if this is helpful, but maybe it answers your question (?):

> a configuration that would result in a single input producing output in both formats but couldn't make it work 

The idea I had in mind is maybe to reformat the instruction `html-math-method` to `katex` once the `pseudocode` filter is used, but honestly, this is only a guess.

**Reply by dragonstyle** [reply]
I appreciate this - I hadn't noticed the changes to the math method (and I'm not sure I understand how it affects the pseudocode rendering, if it does)! It's a great clue for me to look into...
